subsys/fs/nvs/nvs.c

```diff
@@ -654,45 +660,104 @@ int nvs_reinit(struct nvs_fs *fs)
 		goto err;
 	}
 
+	if (sectors_in_use == fs->sector_count) {
+		/* one sector should always be empty, unless power was
+		 * cut during garbage collection, clear the last sector and
+		 * start gc again on the entry sector.
+		 */
+		SYS_LOG_DBG("Erasing last sector");
+		addr = fs->entry_sector * fs->sector_size;
+		/* calculate last sector start */
+		_nvs_addr_decrease(fs, &addr, fs->sector_size);
+		/* erase the last sector */
+		rc = _nvs_flash_erase(fs, addr, fs->sector_size);
```

Comment:
I don't understand why always physically the last sector is erased?

Reply:
Hi Andrzej, the last added sector is erased. When all sectors are used this should be just before the entry sector.

Reply:
NVS uses N sectors, numbered form 0 to N-1. NVS stores entries in N-1 sectors, one sector should be always the scratch sectors.
Why once we got this "GC power down failure" you assume that it was the N-1nd is the properer scratch sector, not any else sector sector are used in round-robin fashion?

Reply:
@lemrey ^^ Did I miss something?

Reply:
@nvlsianpu, NVS always keeps an empty sector, this sector is always just before the entry sector. When gc is started this empty sector is used to copy data from the entry sector. Now what happens if gc is incomplete: all sectors will be in use. We can find the entry sector by looking at the id's. As gc is always done on the entry sector we now what sector gc was doing, we also know where the data was going: to the latest added sector (which is just before the entry sector if all sectors are in use).
Maybe it's because my comment should be "erase last added sector" instead of "erase last sector".

Reply:
My understanding is that we are always erasing the entry sector, but the entry sector moves forward so we are actually cyclying through all the sectors as one would expect. If there is no scratch area then the sector before the entry sector was the one used as scratch area, because an entry sector becomes the scratch area when the entry sector is moved forward.

Reply:
OK, it was my misunderstanding, because I forget and misinterpreted the  meaning of fs->entry_sector field.